### PR TITLE
Update trace tooling

### PR DESCRIPTION
**Pull request by ctubbsii**
Remove and simplify some trace wrapping internal code that made it
harder to migrate to htrace4 or complicated deprecating htrace
(whichever we end up choosing)

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java [hunk 1]
```diff
@@ -138,12 +140,12 @@ public void startDebugLogging() {
 
   public void startTracing(String applicationName) {
     if (trace) {
-      Trace.on(applicationName);
+      Trace.startSpan(applicationName, Sampler.ALWAYS);
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java]: I'm not sure if "ALWAYS" is best here, or if the default is better... the default appears to disable the span if tracing is not enabled for the current thread... which seems like what we want, but I'm unsure of the implications.

File: core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java [hunk 2]
```diff
@@ -138,12 +140,12 @@ public void startDebugLogging() {
...
     }
   }
 
   public void stopTracing() {
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java]: This should be inline'd.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftScanner.java [hunk 3]
```diff
@@ -311,14 +311,12 @@ else if (log.isTraceEnabled())
...
-          } finally {
-            locateSpan.stop();
           }
         }
 
-        Span scanLocation = Trace.start("scan:location");
-        scanLocation.data("tserver", loc.tablet_location);
-        try {
+        try (TraceScope scanLocation = Trace.startSpan("scan:location")) {
+          scanLocation.getSpan().addKVAnnotation("tserver".getBytes(UTF_8),
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftScanner.java]: The new API no longer takes strings?

Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftScanner.java]: The old API never took strings... it took `byte[]`. We had a utility class to do this, but it was confusing, so I inline'd it. HTrace 3.2 and later API actually changed it to String, so these will get simpler when I get that far.

File: core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java (hunk 1 above)
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java]: Leaving any changes to sampling to future work.

**Comment by ctubbsii** [comment]
This is ready for review for now. I'd like to merge this in as a checkpoint towards further upgrading htrace to 4.1. This gets us to 3.2, with a bunch of other reliability improvements.

**Comment by milleruntime** [comment]
I am not familiar with the changes between versions but I can take a look at this.

File: server/base/src/main/java/org/apache/accumulo/server/util/ChangeSecret.java [hunk 4]
```diff
@@ -69,19 +70,21 @@ public static void main(String[] args) throws Exception {
     argsList.add("--old");
     argsList.add("--new");
     argsList.addAll(Arrays.asList(args));
-    opts.parseArgs(ChangeSecret.class.getName(), argsList.toArray(new String[0]));
+    try (TraceScope clientSpan = opts.parseArgsAndTrace(ChangeSecret.class.getName(),
```
Comment (milleruntime) [inline server/base/src/main/java/org/apache/accumulo/server/util/ChangeSecret.java]: The clientSpan variable is never used. You could take out the try with resources and call `clientSpan.close();` at the end.  Which is almost the same thing but IDEs and code analysis tools will at least recognize the use of clientSpan.  IntelliJ gives me this warning FYI.  I think the only difference is if and error is thrown the span wouldn't get closed though.

File: core/src/main/java/org/apache/accumulo/core/trace/TraceUtil.java [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+import org.apache.htrace.TraceScope;
+import org.apache.htrace.impl.CountSampler;
+import org.apache.htrace.impl.ProbabilitySampler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility class for tracing within Accumulo. Not intended for client use!
+ */
+public class TraceUtil {
```
Comment (milleruntime) [inline core/src/main/java/org/apache/accumulo/core/trace/TraceUtil.java]: I like this.  Drop all those similar classes and all the static code collected in one place.  Good work.

**Comment by mjwall** [comment]
HTrace has been retired.  I am not sure how well the newer supported Java.  Is this worth the effort?  Could move to Zipkin or Opentracing

**Comment by mjwall** [comment]
haha, @mikewalch is too fast

File: server/base/src/main/java/org/apache/accumulo/server/util/ChangeSecret.java (hunk 4 above)
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/util/ChangeSecret.java]: The design of the API is to use this with try-with-resources. I had to suppress the warning from javac. We could avoid warnings by referencing it by annotating the spans better. But, a lot of our problems throughout our code involved manual closing of these spans.... and trace bugs due to failure to do so. The try-with-resources construct, as the trace API intends, protects against those kinds of errors. Can you suppress this warning type in IntelliJ?

File: core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java [hunk 6]
```diff
@@ -136,21 +139,15 @@ public void startDebugLogging() {
...
-    Trace.off();
+  public TraceScope parseArgsAndTrace(String programName, String[] args, Object... others) {
+    parseArgs(programName, args, others);
+    return trace ? Trace.startSpan(programName, Sampler.ALWAYS) : NullScope.INSTANCE;
   }
 
   @Override
   public void parseArgs(String programName, String[] args, Object... others) {
     super.parseArgs(programName, args, others);
     startDebugLogging();
```
[the remark this answers is not in the document]
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/cli/ClientOpts.java]: The biggest problem is that the client utils were automatically doing tracing, and were responsible for closing/cleaning up, but failed to do so, when the client tool was written without any intention to trace or awareness of tracing. I'm not a fan of having tracing be built in to the argument parsing at all... but this was a cleaner fix than pushing all those cases down into their respective tools.

**Comment by ctubbsii** [comment]
> HTrace has been retired. I am not sure how well the newer supported Java. Is this worth the effort? Could move to Zipkin or Opentracing

I've looked into Opentracing. I don't think it's mature enough at this point. I would prefer to unretire the HTrace project, have it graduate, and add opentracing compatibility to it. For now, it is worth the effort to get in line with what Hadoop 3 is using... which is 4.1.0-incubating. However, getting to 3.2 was aa major checkpoint towards that.

**Pull request merged**